src/meta/api/src/schema_api_impl.rs

```diff
@@ -2251,12 +2253,25 @@ impl<KV: kvapi::KVApi<Error = MetaError> + ?Sized> SchemaApi for KV {
         }
 
         let mut new_table_meta_map: BTreeMap<u64, TableMeta> = BTreeMap::new();
-        for (req, (tb_meta_seq, table_meta)) in update_table_metas.iter().zip(tb_meta_vec.iter()) {
+        for (req, (tb_meta_seq, table_meta)) in
+            update_table_metas.iter_mut().zip(tb_meta_vec.iter())
+        {
             let tbid = TableId {
                 table_id: req.0.table_id,
             };
             // `update_table_meta` MUST NOT modify `shared_by` field
             let table_meta = table_meta.as_ref().unwrap();
+
+            if let Some(virtual_schema) = &mut req.0.new_table_meta.virtual_schema {
+                if virtual_schema.fields.len() > VIRTUAL_COLUMNS_LIMIT {
+                    return Err(KVAppError::AppError(AppError::VirtualColumnTooMany(
+                        VirtualColumnTooMany::new(req.0.table_id, VIRTUAL_COLUMNS_LIMIT),
+                    )));
+                }
+                for virtual_field in virtual_schema.fields.iter_mut() {
+                    virtual_field.data_types.dedup();
```

Comment:
We also need to check that the `column_id` of the virtual field is greater than or equal to `3_000_000_001` to avoid having a virtual column `column_id` that conflicts with a normal column.